[NEW] Integrated GCM and added notifications support

leonardoaramaki opened the pull request:
<!-- INSTRUCTION: Keep the line below to notify all core developers about this new PR -->
@RocketChat/android 

This  PR depends on https://github.com/RocketChat/Rocket.Chat.Kotlin.SDK/pull/33 been merged.

leonardoaramaki committed:
Add google-services.json

leonardoaramaki committed:
Implemente LocalRepository using SharedPrefences

leonardoaramaki committed:
Configure gcm

leonardoaramaki committed:
Implement registering of generated instance id token on login and also on token rotation

leonardoaramaki committed:
Add default notification icon

leonardoaramaki committed:
Implement notifications support

leonardoaramaki committed:
Add VIBRATE permission

rafaelks commented:
Looks like the PR has conflicts @leonardoaramaki.

rafaelks commented on app/src/main/java/chat/rocket/android/authentication/login/presentation/LoginPresenter.kt:
```diff
@@ -53,4 +56,15 @@ class LoginPresenter @Inject constructor(private val view: LoginView,
     }
+
+    private suspend fun registerPushToken() {
+        // TODO: put it on constructor
```
This code could be more generic? I'm sure we'll have to call in different activities later.

leonardoaramaki replied:
I agree this has to be discussed better later since we're using coroutines here we don't have callbacks.

rafaelks commented on app/src/main/java/chat/rocket/android/authentication/twofactor/presentation/TwoFAPresenter.kt:
```diff
@@ -47,4 +50,15 @@ class TwoFAPresenter @Inject constructor(private val view: TwoFAView,
     }
+
+    private suspend fun registerPushToken() {
+        // TODO: put it on constructor
```
There you go... it's already being called twice? This code could definitely be more generic and don't repeated.

leonardoaramaki committed:
Merge branch 'develop-2.x' into notifications

leonardoaramaki committed:
Fixed code to embrace the new -dev flavor. Changes includes a newly
generated google-services.json file including chat.rocket.android.dev
app within clients

leonardoaramaki committed:
Remove app/libs from .gitignore or else CI fails

leonardoaramaki committed:
Fix broken sdk api calls

leonardoaramaki committed:
Point rest urls to 'open'

luciofm commented on app/src/main/java/chat/rocket/android/authentication/login/presentation/LoginPresenter.kt:
```diff
@@ -56,4 +61,15 @@ class LoginPresenter @Inject constructor(private val view: LoginView,
+
+    private suspend fun registerPushToken() {
+        // TODO: put it on constructor
+        val localRepository: LocalRepository = SharedPreferencesRepository(navigator.activity)
```
You should inject the Repository into the Presenter, and the (application)Context can be injected into the repository.

The presenter should never directly know or access any context, Activity or Application

leonardoaramaki replied:
Actually, we should make RocketChatApplication a singleton. But the presenter here is not aware of the context, since we get it from the Navigator which is injected already. Ideally though, we should have the appContext here

luciofm replied:
it is already, just as a `Context`
```
    @Provides
    @Singleton
    fun provideContext(application: Application): Context {
        return application
    }
```

luciofm replied:
But either way, the presenter should never directly see or touch the Context...

luciofm commented on app/src/main/java/chat/rocket/android/authentication/twofactor/presentation/TwoFAPresenter.kt:
```diff
@@ -50,4 +54,15 @@ class TwoFAPresenter @Inject constructor(private val view: TwoFAView,
+
+    private suspend fun registerPushToken() {
+        // TODO: put it on constructor
+        val localRepository: LocalRepository = SharedPreferencesRepository(navigator.activity)
```
Same as on LoginPresenter, also I would create an Interactor for this, so you can inject the same interactor on both presenters and just call a single method on the Interactor.

luciofm commented on app/src/main/java/chat/rocket/android/infrastructure/SharedPreferencesRepository.kt:
```diff
@@ -0,0 +1,21 @@
+import android.content.Context
+import android.content.SharedPreferences
+
+class SharedPreferencesRepository(context: Context) : LocalRepository {
```
on `AppModule` add a 
```
@Inject
fun provideSharedPreferenceRepository(preferences: SharedPreferences): LocalRepository {
    return  SharedPreferencesRepository(preferences)
}
```

and
```
class SharedPreferencesRepository(private val preferences: SharedPreferences) : LocalRepository {
```
  
So you can inject a `LocalRepository` into Interactors

luciofm commented on app/src/main/java/chat/rocket/android/push/PushManager.kt:
```diff
@@ -0,0 +1,604 @@
+            return
+        }
+
+        val lastPushMessage = PushMessage(title, message, image, ejson, count, notId, summaryText, style)
```
why lastPushMessage and not just pushMessage? With that name seems that you are dealing with an old/previous push message

luciofm commented on app/src/main/java/chat/rocket/android/push/di/PushModule.kt:
```diff
@@ -0,0 +1,64 @@
+
+@Module
+@Singleton
+class PushModule(val context: FirebaseTokenService) {
```
Do we need a PushModule, can't it be on the AppModule?

leonardoaramaki committed:
Implement dagger support for service injections

leonardoaramaki committed:
Merge branch 'develop-2.x' into notifications

leonardoaramaki committed:
Set translatable to false on gcm_sender_id
